### DISCUSSION TITLE
Accessing Serial Port on different platforms.

### DISCUSSION THREAD

**Discussion by Avalyah** [dotnet/maui#4526] [Q&A]
Since .NET 6 was supposed to bring the cross-platform API under one hood, I would like to ask whether SerialPort was also considered for that. When moving from a Windows only app on WPF to .NET Maui I was kind of expecting to just use System.IO.Ports.SerialPort for all my serial port needs, however it seems to not be the case. 

With Windows it works well, on Android I had to use UsbManager from Android.Hardware.Usb.UsbManager, but now I would like to also run my app on MacOS. I once tried porting my Winforms app to MacOS using mono and the System.IO.Ports.SerialPort worked out of the box so with .Net Maui II expected the same. However, SerialPort is not supported on MacCatalyst.

How would I access the Serial Ports on a Mac with my .NET MAUI app? Is that at all possible?

**Comment by Rjvs** [comment]
[Steve Troughton-Smith](https://github.com/steventroughtonsmith/CatalystMenuBarStatusWindow) shows how to create a full-blown Mac app that can also be controlled by a Catalyst app (using Swift) and apparently it used to be possible to access native bundles from Xamarin Forms, so it should be possible to access the serial port from native code and in turn access that from .Net.

**Reply by HiberXen** [reply]
is it   running well on android? can you share some demo code? @Avalyah

**Reply by Avalyah** [reply]
I haven't tested it a lot, but it did connect and sent data correctly.  Here is the Connect method:
`   
            Activity act = Platform.CurrentActivity;

            UsbManager manager = (UsbManager)act.GetSystemService(Context.UsbService);

            IDictionary<string, UsbDevice> devicesDictionary = manager.DeviceList;

            UsbDevice dvc = devicesDictionary.ElementAt(0).Value;

            string ACTION_USB_PERMISSION = "rzepak";

            var interf = dvc.GetInterface(1);

            outEndpoint = interf.GetEndpoint(1);

            PendingIntent mPermissionIntent = PendingIntent.GetBroadcast(act, 0, new Intent(ACTION_USB_PERMISSION), 0);
            IntentFilter filter = new IntentFilter(ACTION_USB_PERMISSION);

            if (manager.HasPermission(dvc) == false) manager.RequestPermission(dvc, mPermissionIntent);

            deviceConnection = manager.OpenDevice(dvc);

            if (deviceConnection != null) deviceConnection.ClaimInterface(interf, true).ToString();
            else return false;
`

Here is how I write to it (device connection is an Android.Hardware.Usb.UsbDeviceConnection type, opened in connect method) :

`  
            byte[] test = new byte[] { input };

            if (deviceConnection.BulkTransfer(outEndpoint, test, test.Length, 500) < 0) Connected = false;
`

**Reply by EJoffre** [reply]
Ok for write but for read ?

**Reply by EJoffre** [reply]
@Avalyah ?

**Reply by kp-121** [reply]
@EliseJoffreIUT Which endpoint you use determines the read & write. If you want to read and write you'll need a read endpoint and a write endpoint. See https://developer.android.com/guide/topics/connectivity/usb/host#api

**Reply by EJoffre** [reply]
Thanks,Finally, I preferred to use data processing in bluetooth for Android and iOS and to use USB for windows

**Comment by fontaipi** [comment]
Is `Android.Hardware.Usb` going to be available for  .NET MAUI project ? At the moment, how can I access USB interface of my Android device with .NET MAUI ?

**Reply by Avalyah** [reply]
It is already available. You can include it like this:

using Android.Hardware.Usb;

**Comment by yurkinh** [comment]
Here is another solution with a similar syntax as in MS library
https://github.com/aidenwlf/MauiAndroidSerialPortNative/tree/master
